pyquil/quiltcalibrations.py

```diff
@@ -125,6 +132,46 @@ def fill_placeholders(obj, placeholder_values: Dict[Union[FormalArgument, Parame
         # raise ValueError(f"Unable to fill placeholders in  object {obj}.")
 
 
+def _convert_to_calibration_match(
+    instruction: Union[quil_rs.Gate, quil_rs.Measurement],
+    calibration: Union[quil_rs.Calibration, quil_rs.MeasureCalibrationDefinition],
+) -> Optional[CalibrationMatch]:
+    if calibration is None:
+        return None
+
+    if isinstance(instruction, quil_rs.Gate) and isinstance(calibration, quil_rs.Calibration):
+        target_qubits = instruction.qubits
+        target_values = instruction.parameters
+        parameter_qubits = calibration.qubits
+        parameter_values = calibration.parameters
+        py_calibration = DefCalibration._from_rs_calibration(calibration)
+    elif isinstance(instruction, quil_rs.Measurement) and isinstance(calibration, quil_rs.MeasureCalibrationDefinition):
+        target_qubits = [instruction.qubit]
+        target_values = (
+            [] if not instruction.target else [MemoryReference._from_rs_memory_reference(instruction.target)]
+        )
+        parameter_qubits = [] if not calibration.qubit else [calibration.qubit]
+        parameter_values = [MemoryReference._from_parameter_str(calibration.parameter)]
+        py_calibration = DefMeasureCalibration._from_rs_measure_calibration_definition(calibration)
+    else:
+        return None
```

Comment:
I would have thought more of this logic would be delegated to `quil-rs` and this function body would be just the FFI call and maybe class transformations. Does that not work out though?

Reply:
Yeah, it is an unfortunate amount of logic. This really is all just class transformations and massaging data into a `pyQuil` friendly format.

`pyQuil`s `match_calibration` and `Program.match_calibrations` return a `CalibrationMatch` which is defined as:

```py
@dataclass
class CalibrationMatch:
    cal: Union[DefCalibration, DefMeasureCalibration]
    settings: Dict[Union[FormalArgument, Parameter], Any]
```

`quil_rs` has no such construct, it just returns a matching calibration definition, so we have to massage the calibration and instruction into a `CalibrationMatch`

`Program.get_calibration` and `Program.calibrate` are quite a bit cleaner because they return calibration definitions, and instructions, respectively. Those probably look more like what you were expecting to see.